cairo/src/utils/signature.cairo

```diff
@@ -84,4 +86,49 @@ namespace Signature {
         let (public_key_point) = ec_add(minus_point1, point2);
         return (public_key_point=public_key_point, success=1);
     }
+
+    // @notice Recovers the Ethereum address from a signature.
+    // @dev If the public key point is not on the curve, the function returns success=0.
+    // @dev: This function does not validate the r, s values.
+    // @param msg_hash The signed message hash.
+    // @param r The r value of the signature.
+    // @param s The s value of the signature.
+    // @param y_parity The y parity value of the signature. true if odd, false if even.
+    // @return The Ethereum address.
+    func try_recover_eth_address{
+        range_check_ptr, bitwise_ptr: BitwiseBuiltin*, keccak_ptr: KeccakBuiltin*
+    }(msg_hash: BigInt3, r: BigInt3, s: BigInt3, y_parity: felt) -> (success: felt, address: felt) {
+        alloc_locals;
+        let (public_key_point, success) = try_recover_public_key(
+            msg_hash=msg_hash, r=r, s=s, y_parity=y_parity
+        );
+        if (success == 0) {
+            return (success=0, address=0);
+        }
+        let (x_uint256) = bigint_to_uint256(public_key_point.x);
+        let (y_uint256) = bigint_to_uint256(public_key_point.y);
+        let address = Internals.public_key_point_to_eth_address(x=x_uint256, y=y_uint256);
+        return (success=success, address=address);
+    }
+}
+
+namespace Internals {
+    // @notice Converts a public key point to the corresponding Ethereum address.
+    // @param x The x coordinate of the public key point.
+    // @param y The y coordinate of the public key point.
+    // @return The Ethereum address.
+    func public_key_point_to_eth_address{
+        range_check_ptr, bitwise_ptr: BitwiseBuiltin*, keccak_ptr: KeccakBuiltin*
+    }(x: Uint256, y: Uint256) -> felt {
+        alloc_locals;
+        let (local elements: Uint256*) = alloc();
+        assert elements[0] = x;
+        assert elements[1] = y;
+        let (point_hash: Uint256) = keccak_uint256s_bigend(n_elements=2, elements=elements);
```

Comment:
do we need `bigend`?

Reply:
does it depend on the initial endianness of the data?

Reply:
yes, it depends on the endianness of the data. In that case, the data is converted to little-endian to be fed to the keccak builtin, with:

```
// Serializes a uint256 number in a keccak compatible way.
// The argument 'bigend' is either 0 or 1, representing the endianness of the given number.
func keccak_add_uint256{range_check_ptr, bitwise_ptr: BitwiseBuiltin*, inputs: felt*}(
    num: Uint256, bigend: felt
) {
    if (bigend != 0) {
        let (num_reversed) = uint256_reverse_endian(num=num);
        tempvar bitwise_ptr = bitwise_ptr;
        tempvar high = num_reversed.high;
        tempvar low = num_reversed.low;
    } else {
        tempvar bitwise_ptr = bitwise_ptr;
        tempvar high = num.high;
        tempvar low = num.low;
    }

    %{
        segments.write_arg(ids.inputs, [ids.low % 2 ** 64, ids.low // 2 ** 64])
        segments.write_arg(ids.inputs + 2, [ids.high % 2 ** 64, ids.high // 2 ** 64])
    %}

    assert inputs[1] * 2 ** 64 + inputs[0] = low;
    assert inputs[3] * 2 ** 64 + inputs[2] = high;

    let inputs = inputs + 4;
    return ();
}
```